File: web/pages/devices.ts
import { api, copyText, escapeHTML, qs, relative } from "../api";
import { onRelayEvent, relayRequest } from "../events";
import { navigate } from "../navigation";
import type { Device, RemoteProcess, Status } from "../types";

function processState(process: RemoteProcess) {
  if (process.status === "starting") return "STARTING";
  if (process.status === "running") return "RUNNING";
  if (process.status === "lost") return "LOST";
  return process.signal || `EXIT ${process.exit_code ?? "?"}`;
}

function deviceRows(devices: Device[]) {
  return devices.length ? devices.map(device => `<a class="data-row" href="/devices/${device.id}">
    <div><strong>${escapeHTML(device.name)}</strong><div class="meta">${escapeHTML(device.workspace_name)} · ${escapeHTML(device.platform.toUpperCase())}/${escapeHTML(device.arch)}</div></div>
    <span class="status ${device.online ? "online" : ""}">${device.online ? "ONLINE" : `SEEN ${relative(device.last_seen)}`}</span>
  </a>`).join("") : '<p class="empty-state">No devices yet. Enroll one from a workspace.</p>';
}

function processRows(deviceId: string, processes: RemoteProcess[]) {
  return processes.length ? processes.map(process => `<a class="data-row process-row" href="/devices/${deviceId}/processes/${process.id}">
    <div><strong class="mono">${escapeHTML(process.command)}</strong><div class="meta">${escapeHTML(process.cwd || "~")} · ${relative(process.created_at)}</div></div>
    <span class="status ${process.status === "running" ? "online" : ""}">${escapeHTML(processState(process))}</span>
  </a>`).join("") : '<p class="empty-state">No processes yet.</p>';
}

export async function renderDevices() {
  const { devices } = await api<{ devices: Device[] }>("/api/v1/devices");
  qs<HTMLElement>("#page").innerHTML = `<div class="page">
    <header class="page-header"><div><p class="eyebrow">DEVICES</p><h1>Devices</h1></div></header>
    <div id="device-list" class="data-list">${deviceRows(devices)}</div>
  </div>`;
  const refresh = async () => {
    const data = await api<{ devices: Device[] }>("/api/v1/devices");
    qs<HTMLElement>("#device-list").innerHTML = deviceRows(data.devices);
  };
  return onRelayEvent(event => {
    if (event.kind === "relay.connected" || event.kind.startsWith("device.")) void refresh();
  });
}

export async function renderDevice(deviceId: string) {
  const [{ device }, { processes }, relay] = await Promise.all([
    api<{ device: Device }>(`/api/v1/devices/${deviceId}`),
    api<{ processes: RemoteProcess[] }>(`/api/v1/devices/${deviceId}/processes`),
    api<Status>("/api/v1/status"),
  ]);
  const supportsProcess = device.capabilities.includes("process");
  const supportsUpdate = device.capabilities.includes("update");
  const updateCommand = "curl -fsSL https://relay.ohrats.party/install.sh | sh";
  let updateInFlight = false;

  qs<HTMLElement>("#page").innerHTML = `<div class="page">
    <header class="page-header device-header">
      <div><p class="eyebrow">DEVICE</p><h1>${escapeHTML(device.name)}</h1><p class="meta"><a href="/workspaces/${device.workspace_id}">${escapeHTML(device.workspace_name)}</a> · ${escapeHTML(device.platform.toUpperCase())}/${escapeHTML(device.arch)}</p></div>
      <span id="device-status" class="status ${device.online ? "online" : ""}">${device.online ? "ONLINE" : `LAST SEEN ${relative(device.last_seen)}`}</span>
    </header>

    <section class="device-summary">
      <dl class="facts">
        <div><dt>HOST</dt><dd>${escapeHTML(device.hostname)}</dd></div>
        <div><dt>NODE</dt><dd id="node-agent">${escapeHTML(device.agent_version)}</dd></div>
        <div><dt>RELAY</dt><dd>${escapeHTML(relay.version)}</dd></div>
        <div><dt>CAPABILITIES</dt><dd id="node-capabilities">${device.capabilities.length ? device.capabilities.map(value => escapeHTML(value.toUpperCase())).join(" · ") : "NONE"}</dd></div>
      </dl>
      <div class="node-actions">
        ${supportsUpdate
          ? '<button id="update-node" class="text-action" type="button">UPDATE NODE</button>'
          : '<button id="copy-update" class="text-action" type="button">COPY UPDATE COMMAND</button>'}
        <p id="update-state" class="meta">${supportsUpdate ? "Stops active processes while restarting." : "This node predates remote updates."}</p>
        ${supportsUpdate ? "" : `<code class="inline-code">${escapeHTML(updateCommand)}</code>`}
      </div>
    </section>

    <section class="content-section process-launch-section">
      <div class="section-heading"><div><p class="eyebrow">NEW PROCESS</p><h2>Start a PTY</h2></div></div>
      <form id="process-launch" class="process-form">
        <label class="cwd-field">Working directory<input id="process-cwd" spellcheck="false" placeholder="~"></label>
        <label class="command-field">Command<input id="process-command" spellcheck="false" value="sh" required></label>
        <button class="primary-button" type="submit" ${device.online && supportsProcess ? "" : "disabled"}>START</button>
      </form>
      <p id="process-error" class="error">${supportsProcess ? "" : "Update this node to start PTY processes."}</p>
    </section>

    <section class="content-section">
      <div class="section-heading"><div><p class="eyebrow">PROCESSES</p><h2>History</h2></div></div>
      <div id="process-list" class="data-list">${processRows(deviceId, processes)}</div>
    </section>
    ${device.role === "owner" || device.role === "member" ? `<section class="content-section danger-section"><p class="eyebrow">DEVICE</p><a class="text-action danger-text" href="/devices/${deviceId}/delete">REMOVE DEVICE</a></section>` : ""}
  </div>`;

  async function refresh() {
    const [{ device: current }, { processes: currentProcesses }] = await Promise.all([
      api<{ device: Device }>(`/api/v1/devices/${deviceId}`),
      api<{ processes: RemoteProcess[] }>(`/api/v1/devices/${deviceId}/processes`),
    ]);
    const status = qs<HTMLElement>("#device-status");
    status.classList.toggle("online", current.online);
    status.textContent = current.online ? "ONLINE" : `LAST SEEN ${relative(current.last_seen)}`;
    qs<HTMLElement>("#node-agent").textContent = current.agent_version;
    qs<HTMLElement>("#node-capabilities").textContent = current.capabilities.length ? current.capabilities.map(value => value.toUpperCase()).join(" · ") : "NONE";
    qs<HTMLButtonElement>('#process-launch button[type="submit"]').disabled = !current.online || !current.capabilities.includes("process");
    const updateButton = document.querySelector<HTMLButtonElement>("#update-node");
    if (updateButton && !updateInFlight) updateButton.disabled = !current.online || !current.capabilities.includes("update");
    qs<HTMLElement>("#process-list").innerHTML = processRows(deviceId, currentProcesses);
  }

  qs<HTMLFormElement>("#process-launch").addEventListener("submit", async event => {
    event.preventDefault();
    qs<HTMLElement>("#process-error").textContent = "";
    const command = qs<HTMLInputElement>("#process-command").value.trim();
    const cwd = qs<HTMLInputElement>("#process-cwd").value.trim();
    try {
      const result = await relayRequest<{ processId: string }>("process.start", { deviceId, command, cwd, cols: 100, rows: 30 });
      await navigate(`/devices/${deviceId}/processes/${result.processId}`);
    } catch (error) { qs<HTMLElement>("#process-error").textContent = error instanceof Error ? error.message : String(error); }
  });

  document.querySelector<HTMLButtonElement>("#update-node")?.addEventListener("click", async event => {
    const button = event.currentTarget as HTMLButtonElement; button.disabled = true; updateInFlight = true;
    qs<HTMLElement>("#update-state").textContent = "Starting update…";
    try {
      await relayRequest("node.update", { deviceId });
      qs<HTMLElement>("#update-state").textContent = "Updating and restarting…";
    } catch (error) {
      qs<HTMLElement>("#update-state").textContent = error instanceof Error ? error.message : String(error);
      button.disabled = false; updateInFlight = false;
    }
  });
  document.querySelector<HTMLButtonElement>("#copy-update")?.addEventListener("click", event => copyText(updateCommand, event.currentTarget as HTMLButtonElement));

  return onRelayEvent(event => {
    if (event.kind === "relay.connected") { void refresh(); return; }
    if (event.deviceId !== deviceId) return;
    if (event.kind === "node.update.error") {
      qs<HTMLElement>("#update-state").textContent = String(event.detail?.error || "Update failed.");
      updateInFlight = false;
      void refresh();
      return;
    }
    if (event.kind === "node.update.ready") { qs<HTMLElement>("#update-state").textContent = "Restarting node…"; return; }
    if (event.kind === "device.online" && updateInFlight) {
      updateInFlight = false;
      qs<HTMLElement>("#update-state").textContent = "Node updated.";
      void refresh();
      return;
    }
    if (event.kind.startsWith("device.") || event.kind.startsWith("process.")) void refresh();
  });
}

export async function renderDeleteDevice(deviceId: string) {
  const { device } = await api<{ device: Device }>(`/api/v1/devices/${deviceId}`);
  qs<HTMLElement>("#page").innerHTML = `<div class="page narrow-form-page">
    <header class="page-header"><div><p class="eyebrow">${escapeHTML(device.workspace_name.toUpperCase())}</p><h1>Remove device</h1></div></header>
    <section class="content-section danger-section">
      <p>Remove <strong>${escapeHTML(device.name)}</strong> from this workspace?</p>
      <p class="page-copy">The node is disconnected and its process history is deleted.</p>
      <div class="actions"><button id="confirm-device-delete" class="text-action danger-text" type="button">REMOVE</button><a class="text-action" href="/devices/${deviceId}">CANCEL</a></div>
      <p id="delete-error" class="error"></p>
    </section>
  </div>`;
  qs<HTMLButtonElement>("#confirm-device-delete").addEventListener("click", async () => {
    try {
      await api(`/api/v1/devices/${deviceId}`, { method: "DELETE", body: "{}" });
      await navigate(`/workspaces/${device.workspace_id}`);
    } catch (error) {
      qs<HTMLElement>("#delete-error").textContent = error instanceof Error ? error.message : String(error);
    }
  });
}

File: crates/rc-server/src/page_html/devices.rs
use super::{PageContext, authenticated_document, bool_value, esc, format, integer, string};

pub fn devices(context: &PageContext) -> String {
    let rows = if context.devices.is_empty() {
        "<p class=\"empty-state\">No devices yet. Enroll one from a workspace.</p>".into()
    } else {
        context.devices.iter().map(device_row).collect::<String>()
    };
    let can_enroll = context
        .workspaces
        .iter()
        .any(|workspace| string(workspace, "role") == "owner");
    let body = format!(
        "<div class=\"page\" data-live-page=\"devices\"><header class=\"page-header\"><div><h1>Devices</h1></div>{}</header><div class=\"data-list\" id=\"device-list\">{}</div></div>",
        if can_enroll {
            "<a class=\"header-icon-button\" href=\"/devices/enroll\" aria-label=\"Enroll device\" title=\"Enroll device\"><span class=\"ui-icon icon-plus\"></span></a>"
        } else {
            ""
        },
        rows
    );
    authenticated_document(context, "Devices", body, &["live"], &[])
}

fn device_row(device: &serde_json::Value) -> String {
    let id = string(device, "id");
    let name = string(device, "name");
    let platform = string(device, "platform");
    let active = integer(device, "active_processes");
    let online = bool_value(device, "online");
    let status = if online {
        "ONLINE".into()
    } else {
        format!(
            "SEEN {}",
            format::relative(device.get("last_seen").and_then(serde_json::Value::as_i64))
        )
    };
    format!(
        "<a class=\"data-row\" href=\"/devices/{}\" data-device-row=\"{}\"><div class=\"device-row-main\"><span class=\"ui-icon device-platform-icon {}\"></span><div><strong>{}</strong><div class=\"meta\">{} · {}/{}{} </div></div></div><span class=\"status{}\" data-device-status=\"{}\">{}</span></a>",
        esc(&id),
        esc(&id),
        format::platform_icon(&platform),
        esc(&name),
        esc(&string(device, "workspace_name")),
        esc(&platform.to_ascii_uppercase()),
        esc(&string(device, "arch")),
        if active > 0 {
            format!(" · {active} ACTIVE")
        } else {
            String::new()
        },
        if online { " online" } else { "" },
        esc(&id),
        esc(&status)
    )
}

pub fn device(
    context: &PageContext,
    device: &serde_json::Value,
    processes: &[serde_json::Value],
) -> String {
    let id = string(device, "id");
    let name = string(device, "name");
    let role = string(device, "role");
    let platform = string(device, "platform");
    let online = bool_value(device, "online");
    let supports_process = device
        .get("capabilities")
        .and_then(serde_json::Value::as_array)
        .is_some_and(|values| values.iter().any(|value| value.as_str() == Some("process")));
    let can_operate = matches!(role.as_str(), "owner" | "operator");
    let can_manage = role == "owner";
    let process_rows: String = if role == "viewer" {
        "<p class=\"empty-state\">Process history is available to operators and owners.</p>".into()
    } else if processes.is_empty() {
        "<p class=\"empty-state\">No processes yet.</p>".into()
    } else {
        processes
            .iter()
            .map(|process| process_row(&id, process))
            .collect()
    };
    let rename = if can_manage {
        format!(
            "<form method=\"post\" action=\"/devices/{}/rename\" hidden data-device-title-form><input class=\"device-title-input\" name=\"name\" value=\"{}\" aria-label=\"Device name\" required maxlength=\"120\"><input type=\"hidden\" name=\"next\" value=\"/devices/{}\"></form><button class=\"header-icon-button\" type=\"button\" data-device-title-rename aria-label=\"Rename device\" title=\"Rename device\"><span class=\"ui-icon icon-pencil\"></span></button>",
            esc(&id),
            esc(&name),
            esc(&id)
        )
    } else {
        String::new()
    };
    let delete = if can_manage {
        format!(
            "<button class=\"header-icon-button danger-icon-button\" type=\"button\" aria-label=\"Delete {}\" title=\"Delete device\" data-delete-kind=\"device\" data-delete-name=\"{}\" data-delete-description=\"Deletes from RC immediately. If the Node is offline, it clears its old enrollment the next time it connects.\" data-delete-endpoint=\"/api/v1/devices/{}\" data-delete-redirect=\"/devices\"><span class=\"ui-icon icon-trash\"></span></button>",
            esc(&name),
            esc(&name),
            esc(&id)
        )
    } else {
        String::new()
    };
    let terminal = if can_operate {
        format!(
            "<button id=\"open-terminal\" class=\"device-terminal-button\" type=\"button\" aria-label=\"Open terminal\" title=\"Open terminal\" {}><span class=\"ui-icon icon-terminal\"></span></button>",
            if online && supports_process {
                ""
            } else {
                "disabled"
            }
        )
    } else {
        String::new()
    };
    let body = format!(
        "<div class=\"page\" data-device-page=\"{}\" data-supports-process=\"{}\"><section class=\"device-overview\"><header class=\"page-header device-header\"><div><p class=\"eyebrow\">DEVICE</p><div class=\"page-title-row device-title-row\"><span class=\"ui-icon device-platform-icon device-title-platform {}\"></span><h1 data-device-title-view>{}</h1>{}</div><p class=\"error device-title-error\" data-device-title-error></p><p class=\"meta\">{} · {}/{}</p></div><div class=\"device-header-actions\"><span id=\"device-status\" class=\"status{}\">{}</span>{}{}</div></header><dl class=\"facts\"><div><dt>HOST</dt><dd>{}</dd></div><div><dt>NODE VERSION</dt><dd id=\"node-version\">{}</dd></div></dl><p id=\"process-error\" class=\"error\">{}</p></section><section class=\"content-section\"><div class=\"section-heading\"><div><div class=\"badge-container\"><span class=\"badge-text\">01 Processes</span><div class=\"badge-line\"></div></div><h2>History</h2></div></div><div class=\"data-list\" id=\"process-list\">{}</div></section></div>",
        esc(&id),
        supports_process,
        format::platform_icon(&platform),
        esc(&name),
        rename,
        esc(&string(device, "workspace_name")),
        esc(&platform.to_ascii_uppercase()),
        esc(&string(device, "arch")),
        if online { " online" } else { "" },
        if online { "ONLINE" } else { "OFFLINE" },
        terminal,
        delete,
        esc(&string(device, "hostname")),
        esc(&string(device, "version")),
        if supports_process {
            ""
        } else {
            "This RC Node is too old for terminals. Run rc update on the device."
        },
        process_rows
    );
    authenticated_document(context, &name, body, &["live", "device"], &[])
}

fn process_row(device_id: &str, process: &serde_json::Value) -> String {
    let id = string(process, "id");
    let status = string(process, "status");
    format!(
        "<a class=\"data-row process-row\" href=\"/devices/{}/processes/{}\" data-process-row=\"{}\"><div><strong class=\"mono\">{}</strong><div class=\"meta\">{} · {} · {}</div></div><span class=\"status{}\" data-process-status=\"{}\">{}</span></a>",
        esc(device_id),
        esc(&id),
        esc(&id),
        esc(&format::process_label(process)),
        esc(&string(process, "origin").to_ascii_uppercase()),
        esc(&string(process, "created_by_name")),
        esc(&format::relative(
            process
                .get("created_at")
                .and_then(serde_json::Value::as_i64)
        )),
        if status == "running" { " online" } else { "" },
        esc(&id),
        esc(&format::process_state(process))
    )
}

pub fn process(
    context: &PageContext,
    device: &serde_json::Value,
    process: &serde_json::Value,
) -> String {
    let id = string(process, "id");
    let device_id = string(device, "id");
    let status = string(process, "status");
    let running = matches!(status.as_str(), "starting" | "running");
    let role = string(device, "role");
    let owner = string(process, "created_by");
    let controllable = role == "owner" || (role == "operator" && owner == context.user.id);
    let direct = matches!(
        string(process, "origin").as_str(),
        "browser" | "cli" | "api"
    );
    let terminal = bool_value(process, "terminal");
    let interactive = direct && terminal && running && controllable;
    let label = format::process_label(process);
    let terminal_body = if interactive {
        format!(
            "<pre id=\"process-transcript\" class=\"terminal-transcript\">Terminal scrollback is retained in RC Node memory while this process is live.</pre><div id=\"terminal-host\" class=\"terminal-host\" hidden></div><div class=\"mobile-terminal-keys\" aria-label=\"Terminal keys\">{}</div>",
            mobile_keys()
        )
    } else {
        format!(
            "<pre class=\"terminal-transcript\">{}</pre>",
            if terminal {
                "Terminal content is retained only in RC Node memory while the process is live."
            } else {
                "Process content is not retained by RC."
            }
        )
    };
    let actions = if interactive {
        "<div id=\"terminal-actions\" class=\"terminal-actions\"><button class=\"text-button\" data-signal=\"INT\" type=\"button\">CTRL-C</button><button class=\"text-button\" data-signal=\"TERM\" type=\"button\">TERM</button><button class=\"text-button\" data-signal=\"KILL\" type=\"button\">KILL</button></div>"
    } else {
        ""
    };
    let message = string(process, "error");
    let message = if message.is_empty() && running && !controllable {
        format!(
            "Live control belongs to {}.",
            string(process, "created_by_name")
        )
    } else {
        message
    };
    let body = format!(
        "<div class=\"page process-page\" data-process-page=\"{}\" data-device-id=\"{}\" data-process-status=\"{}\" data-process-live=\"{}\" data-process-interactive=\"{}\"><header class=\"page-header process-header\"><div><p class=\"eyebrow\"><a href=\"/devices/{}\">{}</a> / PROCESS</p><h1 class=\"mono process-title\">{}</h1><p class=\"meta\">{} · STARTED BY {} · {}</p></div><span id=\"process-state\" class=\"status{}\">{}</span></header><div id=\"terminal-toolbar\" class=\"terminal-toolbar\"><span class=\"terminal-label\">{}/{}{} </span>{}</div>{}<p id=\"process-message\" class=\"meta process-message\">{}</p></div>",
        esc(&id),
        esc(&device_id),
        esc(&status),
        running,
        interactive,
        esc(&device_id),
        esc(&string(device, "name").to_ascii_uppercase()),
        esc(&label),
        esc(&string(process, "origin").to_ascii_uppercase()),
        esc(&string(process, "created_by_name")),
        esc(&format::relative(
            process
                .get("created_at")
                .and_then(serde_json::Value::as_i64)
        )),
        if status == "running" { " online" } else { "" },
        esc(&format::process_state(process)),
        if terminal { "PTY" } else { "PROCESS" },
        esc(&id.chars().take(8).collect::<String>()),
        if interactive {
            " · <span id=\"control-transport\">CONNECTING</span>"
        } else {
            ""
        },
        actions,
        terminal_body,
        esc(&message)
    );
    authenticated_document(
        context,
        &label,
        body,
        if interactive {
            &["process-terminal"]
        } else {
            &[]
        },
        if interactive {
            &["process-terminal"]
        } else {
            &[]
        },
    )
}

fn mobile_keys() -> String {
    ["ESC", "CTRL", "ALT", "TAB", "LEFT", "UP", "DOWN", "RIGHT"]
        .iter()
        .map(|key| {
            format!(
                "<button type=\"button\" data-terminal-key=\"{}\">{}</button>",
                key,
                match *key {
                    "LEFT" => "←",
                    "UP" => "↑",
                    "DOWN" => "↓",
                    "RIGHT" => "→",
                    value => value,
                }
            )
        })
        .collect()
}
